### PR TITLE
Classlist: add option to specify external reader

### PR DESCRIPTION
**Pull request by sbesson**
See https://trello.com/c/EPqD1xkr/163-externalize-slidebook6reader and https://trello.com/c/k1RljMtO/162-classlist-documentation. This PR is meant to supersede https://github.com/openmicroscopy/bioformats/pull/2440

Follow-up of https://github.com/openmicroscopy/bioformats/pull/2457 which implemented some API to pass per-reader options as key/value pairs, this PR:
- extends the top level Javadoc of the ClassList API to describe the expected format of the configuration file with an example
- adds a whitelisted key `type` to `ClassList` and implements `isWhitelistedKey()` with unit tests
- implements the `type=external` option for a given class to prevent logging missing classes at the error level
- updates the readers.txt configure file to define `SlideBook6Reader` as an external reader using the key/value pair defined above
- removes the `SlideBook6Reader` class and all references to it from the Bio-Formats codebase

Testing:
- all unit and automated tests should remain green with this PR included
- the native Slidebook reader developed by @Intelligent-Imaging should remain functional with this. Note the latest version of the JAR should be functional on Windows, OSX and Linux. Download the latest JAR from http://sites.imagej.net/SlideBook and test minimally the three following configuration 
   - OSX/command-line tools: add the JAR to the classpath using `BF_CP=/path/to/SlideBook6Reader.jar` 
   - Windows/Fiji: add the 3i Jar under `jars` on a fresh Fiji distribution
   - Linux/OMERO: using a merge OMERO 5.3 server, add the JAR under `lib/server` and `lib/client`
   - test the opening/importing functionality using a Slidebook file known to fail with the OME Slidebook reader e.g. https://trac.openmicroscopy.org/ome/ticket/12406
- without the Slidebook Jar, check no ERROR/WARN log level appears due to the missing class

### CODE REVIEW

**Comment by joshmoore** [comment]
Longish-term: what would be the strategy for adding a key with multiple values?

**Comment by sbesson** [comment]
I have not thought about this use case. The obvious and directly breaking change would be to redefine the options as a `Map<String, List<String>>` but this might make the API too complex. Is this a case you envisioned in the case of the key/value pairs added to  `MetadataOptions` on the `metadata` branch?

If we are not sure of the long-term impact of declaring a `type=<type>` option key, I can certainly modify this PR and use `external=true` as I had initially planned.

**Comment by hflynn** [comment]
Mention of SlideBook6Reader needs to be removed from format-pages.txt to fix the broken link in the docs build. We also need to update the notes on this page and remove the metadata page for that reader but those can be done on a separate PR if preferred.

File: docs/sphinx/formats/dataset-table.txt [hunk 1]
```diff
@@ -398,12 +398,12 @@ to open/import a dataset in a particular format.
    * - SM Camera
      - (no extension)
      - Single file
+   * - SPC FIFO Data
```
Comment (hflynn) [inline docs/sphinx/formats/dataset-table.txt]: Thanks for spotting that I forgot to add this! We should probably add the .set extension to the formats-pages.txt too, it isn't on the supported formats table currently

**Comment by hflynn** [comment]
Docs build is green again now thanks!

We could probably remove the 'we wants' from the slidebook page now too but that's a minor nitpick I can sort (as is my comment above). So looks good from the docs perspective.

File: docs/sphinx/formats/dataset-table.txt (hunk 1 above)
Comment (hflynn) [inline docs/sphinx/formats/dataset-table.txt]: Ah, I see, it isn't that I forgot it, it's that I updated the reader but didn't regen the dataset structure table.

**Comment by pwalczysko** [comment]
> - OSX/command-line tools: add the JAR to the classpath using BF_CP=/path/to/SlideBook6Reader.jar
- Windows/Fiji: add the 3i Jar under jars on a fresh Fiji distribution

These two testing points were done exaclty as described in the header of this PR. 
Note that for the BF_CP on OSX and bftools test you have to also do ``export BF_CP``.
Note that you have to change all the relevant jars under Fiji jars folder for the jars from Bioformats-dev-merge in order to make this happen. Also, you have to enable updating site `` http://sites.imagej.net/SlideBook `` in your Fiji's ``Manage Update Sites`` menu and run the Fiji updater, then restart Fiji.
The two tests described in this comment fully passed.

The only ToDo here left is to check the third bullet point test from the header of this PR (.....Linux/OMERO: using a merge OMERO 5.3 server,....). The reason why this test was not done here today is that the imports to omero-merge servers do not work because of another PR https://github.com/openmicroscopy/openmicroscopy/pull/4753.

**Comment by jburel** [comment]
Discussed with @sbesson: having a template on how to update your desktop will be useful so users know how, for example, they can add the new slidebook.jar and import from insight
similar to https://github.com/openmicroscopy/bioformats/pull/2496 http://help.openmicroscopy.org/client-memory.html

It seems that we have an issue at the thumbnail generation:
``
java.lang.AssertionError: File not open.
        at loci.formats.in.SlideBook6Reader.readImagePlaneBuf(Native Method) ~[SlideBook6Reader.jar:5.2.0-SNAPSHOT]
        at loci.formats.in.SlideBook6Reader.openBytes(SlideBook6Reader.java:148) ~[SlideBook6Reader.jar:5.2.0-SNAPSHOT]
        at loci.formats.ImageReader.openBytes(ImageReader.java:453) ~[formats-api.jar:5.2.0-SNAPSHOT]
        at loci.formats.ChannelFiller.openBytes(ChannelFiller.java:156) ~[formats-bsd.jar:5.2.0-SNAPSHOT]
        at loci.formats.ChannelSeparator.openBytes(ChannelSeparator.java:225) ~[formats-bsd.jar:5.2.0-SNAPSHOT]
        at loci.formats.ChannelSeparator.openBytes(ChannelSeparator.java:147) ~[formats-bsd.jar:5.2.0-SNAPSHOT]
        at loci.formats.ReaderWrapper.openBytes(ReaderWrapper.java:342) ~[formats-api.jar:5.2.0-SNAPSHOT]
        at ome.io.bioformats.BfPixelsWrapper.getWholePlane(BfPixelsWrapper.java:359) ~[romio.jar:na]
        at ome.io.bioformats.BfPixelsWrapper.getPlane(BfPixelsWrapper.java:256) ~[romio.jar:na]
        at ome.io.bioformats.BfPixelBuffer.getPlane(BfPixelBuffer.java:180) ~[romio.jar:na]
        at omeis.providers.re.data.PlaneFactory.createPlane(PlaneFactory.java:201) ~[rendering.jar:na]
        at omeis.providers.re.HSBStrategy.getWavelengthData(HSBStrategy.java:103) ~[rendering.jar:na]
        at omeis.providers.re.HSBStrategy.makeRenderingTasks(HSBStrategy.java:217) ~[rendering.jar:na]
        at omeis.providers.re.HSBStrategy.render(HSBStrategy.java:312) ~[rendering.jar:na]
``

**Comment by sbesson** [comment]
Discussed with @jburel, @hflynn and @melissalinkert, merging this as it is as it passes all the Bio-Formats tests and all functionalities are preserved. This means I will close https://github.com/openmicroscopy/bioformats/pull/2440 and 3i now maintains both the reader and the libraries in their provided JAR from 5.2.0 on onwards.

On the OMERO side, the workflow works up to import state but images fail to render when trying to retrieve individual planes as described in https://github.com/openmicroscopy/bioformats/pull/2496#issuecomment-236554783. This issue is related to the use of the `Memoizer` in the stack as some file handles might not be freed/reinitialized in the native libraries. A workaround on our side would be to disable memoization when the native Slidebook reader is used. But this might be something 3i want to investigate and fix in their reader directly. @RichardMyers: let us know if you want more information about the testing instructions to reproduce this issue (minimally it can be done via Bio-Formats command-line tools directly using the `-cache` option).

**Comment by RichardMyers** [comment]
Thanks @sbesson and everyone who made this possible! I am glad to see this change will be in 5.2.0.  I'll investigate the OMERO issue, but I would appreciate some tips on testing.   It does seems likely file handle are remaining open between plane reads.

**Comment by sbesson** [comment]
@RichardMyers: to test and investigate this particular issue, you should be able to use Bio-Formats alone as mentioned in my previous comment. The critical part to reproduce the stack trace is to wrap the reader with a `Memoizer` reader wrapper which serializes and caches the initialized reader as part of the `setId()` call. This wrapper is heavily used by OMERO in particular.

The error reported above shows up when the reader is re-initialized from a cached memoization file and the planes are read. The memoization functionality is not exposed in the graphical ImageJ plugin although it should be possible to use a macro. Using the command-line tools, the following scenario should reproduce the issue:

```
# Will initialize the reader, cache the reader on disk and open the pixels
BF_CP=/path/to/Slidebook.jar ./showinf -cache /path/to/file.sld 
# Will initialize the reader using the file on disk and fail when opening the plane
BF_CP=/path/to/Slidebook.jar ./showinf -cache /path/to/file.sld
```

If you also want to investigate the testing your JARs in an OMERO environment now or later in your development process, we can also point you at some of the resources we have in place for testing OMERO installations with third-party extensions like OMERO.web apps and describe how one would adapt these to test third-party Bio-Formats readers /cc @jburel

**Comment by RichardMyers** [comment]
@sbesson, thanks for the repo steps.  I was able to find/fix the problem.   The SlideBook6Reader plugin will now open the .sld file at the beginning of each method (initFile, openBytes), and close it at the end.  My tests with showinf confirm this fixes the problem.  We'll want to test on OMERO too, but that's going to have to wait a bit longer.

**Comment by jburel** [comment]
@RichardMyers: is the version of the jar 2016-08-10, the latest?
If this is the case, we could test it tomorrow or Friday and get back to you.

**Comment by RichardMyers** [comment]
@jburel, yes I just updated it once yesterday, so 2016-08-10 is correct.

**Comment by jburel** [comment]
@RichardMyers: great, I will run some tests hopefully today

**Comment by jburel** [comment]
@RichardMyers:

SlideBook.jar used version: ``SlideBook6Reader.jar-20160810213131``

Context 1:
 * Running OMERO server on Ubuntu 14.04, Add Slidebook.jar to the server and client repositories. Unable to import any slidebook files (including the files that could be imported prior to the new version). No error found any of the log files. It cannot pass the "reading metadata" stage.

Context 2;
 * Running OMERO server on MacOSX, Add Slidebook.jars to the server and client repositories.
The slidebook files can be imported and can be viewed in OMERO.

We have not tested using a OMERO server running on Windows.

I hope this report will help you.
This PR being closed, we should probably move the discussion to another location, maybe the forum or the mailing lists.

**Comment by ehrenfeu** [comment]
Sorry for jumping in here, but I'm a bit lost :wink: 

As I understand from the thread, the `SlideBook.jar` is *not yet* working with OMERO on Linux, right? Or is this problem specific to Ubuntu 14.04?

Eventually, it is expected to work also in the 5.2 series of OMERO, or would we have to wait for 5.3?

I'm totally amazed by the great work on externalizing the .sld support, very exciting! :smile: 

As @jburel already mentioned moving the discussion somewhere else in the previous comment, feel free to push me away here and relocate this - clearly a github issue will have a *much* smaller audience than a forum / mailing list post...

**Comment by sbesson** [comment]
@ehrenfeu: no worries, using this thread is certainly relevant and completely fine.

This feature (Slidebook externalization) is available as part of Bio-Formats 5.2.0 and above which means 3i provied Slidebook JAR with all platforms support should be available via the Bio-Formats tools (command-line, MATLAB, ImageJ/Fiji) immediately. On the OMERO side, since this release includes breaking model changes, it will only be available with OMERO 5.3.0 and onwards. See also https://github.com/openmicroscopy/ome-help/pull/196 for documentation efforts on how to integrate it.

Regarding Ubuntu vs Linux, I do not have the complete answer but the https://github.com/Intelligent-Imaging/bioformats/issues/3 issue opened by @RichardMyers might be a good place to discuss.

**Comment by ehrenfeu** [comment]
Thanks Sebastien for the heads-up!

**Pull request merged**